gamedaoco/gamedao-protocol
# Scheduling and Campaign settlement

[Figma diagram](https://www.figma.com/file/Jqy2XnQh9ZgvU518Zhp48u/Scheduling-%2B-Campaign-Settlement?node-id=0%3A1)
- Added scheduler, which is responsible for Campaign finalization.
- Updated Flow tests.

**Test scenario:**
- Set runtime: `MaxContributorsProcessing = 4`
- Extrinsic: Create a Campaign
- Extrinsic: Make 10 Contributions
- Ensure: 1st block 4 contributors were finalized
- Ensure: 2nd block 4 contributors were finalized
- Ensure: 3rd block 2 contributors were finalized, and the Campaign was finalized (FAILED or FINALIZED status)

**Do we need to handle this scenario?**
https://github.com/gamedaoco/gamedao-protocol/blob/0cec633f4869d2d43d6a97981d2fe11a86619f7d/flow/src/lib.rs#L818-L824

I've found the following reasons why it can fail:
- from_account has one or multiple BalanceLocks -> frozen balance
- from_account reserved balance was slashed
- to_account was not in DustRemovalWhitelist, and its balance < existential deposit
- to_account ArithmeticError::Overflow

Based on this, it looks like not such a possible scenario.
